✨ feat: Added Support for Flagship Gemini Image (🍌 Nano Banana) models for Image generation and editing

devilb2103 opened the pull request:
# Added Gemini Flash Image Generation Tool (Nano Banana + Nano Banana Pro)

https://github.com/user-attachments/assets/b67fc374-6116-422a-bca9-8641e27f7fc9

## Summary

This PR (addresses https://github.com/danny-avila/LibreChat/issues/9283 and https://github.com/danny-avila/LibreChat/issues/6065) implements a new image generation tool using Google's Gemini 2.5 Flash Image Model with Vertex AI integration. The tool supports both text-only image generation and image context-aware generation for editing/modification requests. It includes handling of rejected tool calls due to content policy violoations, multiple file storage strategies, and seamless integration with LibreChat's existing agent tool ecosystem.

**Key Features:**
- Text-to-image generation using Gemini 2.5 Flash Image Model
- Image context support for editing existing images
- Multiple storage strategies (local, S3, Azure, Firebase)
- Safety filtering with user-friendly error messages

**Configuration Requirements:**
- Google Cloud service account JSON file required at `api/data/auth.json`
- Docker volume mapping: `./api/data/auth.json:/app/api/data/auth.json`
- Vertex AI API permissions for the service account
- LibreChat YAML configuration update required

## Change Type

- [x] New feature (non-breaking change which adds functionality)

## Testing

The feature has been tested for:
- ESLint compliance (all files pass)
- Tool registration and integration
- Docker container compatibility
- File path resolution in containerized environment

### **Test Configuration**:
- **Environment**: Docker containerized deployment
- **Dependencies**: `@google/genai: ^1.17.0` added to `api/package.json`
- **Authentication**: Google Cloud service account with Vertex AI permissions
- **Storage**: Local and S3 file strategy tested (other strategies compatible - follow Open AI Image Tools Logic)

**Test Steps:**
1. Build Docker container with new dependencies
2. Configure Google service account credentials
3. Enable `gemini_image_gen` tool in an agent
4. Test text-to-image generation
5. Test image context-aware generation
6. Verify safety filtering functionality
7. Verify if generated Images show up in attachments drop down in the collapsable LLM Tools section at the right hand side of UI

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes
- [x] Any changes dependent on mine have been merged and published in downstream modules.

## Files Modified

### Core Implementation
- `api/app/clients/tools/structured/GeminiImageGen.js` - Main tool implementation
- `api/package.json` - Added `@google/genai` dependency
- `packages/data-provider/src/config.ts` - Added to `imageGenTools` set

### Integration
- `api/app/clients/tools/index.js` - Tool export
- `api/app/clients/tools/manifest.json` - Tool manifest entry
- `api/app/clients/tools/util/handleTools.js` - Tool registration and configuration
- `api/server/services/ToolService.js` - Import statement

### Documentation
- `api/app/clients/tools/structured/README-GeminiNanoBanana.md` - Implementation guide

## Breaking Changes

None. This is a purely additive feature that doesn't modify existing functionality.

## Configuration Notes

**Required Docker Configuration:**
```yaml
volumes:
  - ./api/data/auth.json:/app/api/data/auth.json
```

**LibreChat YAML Configuration:**
Add `gemini_image_gen` to the `includedTools` array in your `librechat.yaml`:
```yaml
includedTools: ['gemini_image_gen']
```

## Environment Variables

Add these to your `.env` file to configure the Gemini Image Generation tool:

| Variable | Description | Default | Required |
|----------|-------------|---------|----------|
| `GEMINI_IMAGE_PROVIDER` | Provider to use: `vertex` (service account) or `gemini` (API key) | Auto-detected | No |
| `GEMINI_IMAGE_MODEL` | Model ID for image generation | `gemini-2.5-flash-image-preview` | No |
| `GEMINI_API_KEY` | Gemini API key (required if using `gemini` provider) | - | If provider=gemini |
| `GOOGLE_SERVICE_KEY_FILE` | Path to service account JSON (for Vertex AI) | `api/data/auth.json` | If provider=vertex |
| `GOOGLE_CLOUD_LOCATION` | Google Cloud region for Vertex AI | `global` | No |
| `GOOGLE_APPLICATION_CREDENTIALS` | Vertex AI Service account credentials path | None | Yes |

**Note:** If `GEMINI_IMAGE_PROVIDER` is not set, the tool auto-detects:
> - Uses **Vertex AI** if `GOOGLE_SERVICE_KEY_FILE` exists or `api/data/auth.json` is present
> - Uses **Gemini API** if `GEMINI_API_KEY` is set

**Note:** Service Account Credentials must be manually set in .env in the following manner:
```env
GOOGLE_APPLICATION_CREDENTIALS="data/auth.json"
```

**Google Cloud Setup:**
1. Create a service account with Vertex AI permissions
2. Download the service account key as JSON
3. Place the file at `api/data/auth.json`
4. Ensure the project ID matches your Google Cloud project

devilb2103 committed:
Added fully functioning Agent Tool supporting Google's Nano Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

falkenbt reviewed (commented):
Really nice, I got  this to work on my machine. There is just a small issue with the credentials path, see inline comment.

devilb2103 committed:
🔧 refactor: Update Google credentials handling in GeminiImageGen.js

* Refactored the credentials path to follow a consistent pattern with other Google service integrations, allowing for an environment variable override.
* Updated documentation in README-GeminiNanoBanana.md to reflect the new credentials handling approach and removed references to hardcoded paths.

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 commented on api/app/clients/tools/structured/GeminiImageGen.js:
```diff
@@ -0,0 +1,658 @@
+    }
+
+    // Hardcoded path to auth.json
+    const credentialsPath = path.resolve('./data/auth.json');
```
this should be addressed now

devilb2103 commented on api/app/clients/tools/structured/GeminiImageGen.js:
```diff
@@ -0,0 +1,658 @@
+      const credentialsPath = path.resolve('./data/auth.json');
+
+      // Set environment variables for Vertex AI (required by Google GenAI library)
+      process.env.GOOGLE_CLOUD_PROJECT = 'hingegpt';
```
this has been addressed with the newer commits

lukaswelte commented:
@devilb2103 thanks for that trivial hint - sorry I hadn't checked that 🤦 
Error was in IAM permissions on GCP - problem solved, it works like a charm 🚀

falkenbt reviewed (commented):
Looks good to me now, thanks a lot. I cannot approve though, I'm just a user without elevated access.

devilb2103 commented:
> Why is Vertex AI API used here? This also works [with Gemini API](https://ai.google.dev/gemini-api/docs/image-generation#image_generation_text-to-image)
> 
> ```REST
> 
> curl -s -X POST
> 
>   "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-image-preview:generateContent" \
> 
>   -H "x-goog-api-key: $GEMINI_API_KEY" \
> 
>   -H "Content-Type: application/json" \
> 
>   -d '{
> 
>     "contents": [{
> 
>       "parts": [
> 
>         {"text": "Create a picture of a nano banana dish in a fancy restaurant with a Gemini theme"}
> 
>       ]
> 
>     }]
> 
>   }' \
> 
>   | grep -o '"data": "[^"]*"' \
> 
>   | cut -d'"' -f4 \
> 
>   | base64 --decode > gemini-native-image.png
> 
>   ```
> 
> 

It should be fairly easy to have another librechat manifest var to switch between the already supported VertexAI option with GeminiAPI

This implementation makes it convenient for those who have access to Googles VertexAI tools though

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
🛠️ refactor: Remove unnecessary whitespace in handleTools.js

ZenDevMaster commented:
It must be discouraging for contributors to submit pull requests and then be ignored.

marlonka commented on api/app/clients/tools/structured/GeminiImageGen.js:
```diff
@@ -0,0 +1,657 @@
+
+      // Generate image using Gemini Vertex AI with optional image context
+      apiResponse = await ai.models.generateContent({
+        model: 'gemini-2.5-flash-image-preview',
```
@devilb2103 can you implement it via .env variable to Change the model name?

devilb2103 replied:
addressed

tatwong commented:
> Does it support Gemini 3 Pro Image (aka Nano Banana Pro) already? Would be awesome for admins to choose in env or librechat.yaml the model or have a dropdown in the tool UI: https://docs.cloud.google.com/vertex-ai/generative-ai/docs/models/gemini/3-pro-image

Same question, so is that means I can use Gemini 3 Pro to generate Infographics now on LibreChat too?

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

danny-avila commented on api/app/clients/tools/structured/GeminiImageGen.js:
New modules are not being accepted in commonjs. You must migrate this to the packages/api workspace for benefiting from TypeScript

devilb2103 replied:
addressed this

danny-avila commented on api/package.json:
```diff
@@ -41,6 +41,7 @@
     "@azure/search-documents": "^12.0.0",
     "@azure/storage-blob": "^12.27.0",
     "@google/generative-ai": "^0.24.0",
+    "@google/genai": "^1.17.0",
```
Must also be added as a peer dependency with matching version in `packages/api/package.json`

danny-avila commented on api/app/clients/tools/util/handleTools.js:
```diff
@@ -218,6 +220,38 @@ const loadTools = async ({
+      const authValues = await loadAuthValues({ userId: user, authFields });
+      const imageFiles = options.tool_resources?.[EToolResources.image_edit]?.files ?? [];
+      let toolContext = '';
+      for (let i = 0; i < imageFiles.length; i++) {
```
try to consolidate this into its own function avoid bloating this module, also imported from `packages/api` i.e. `@librechat/api`

usnavy13 commented:
@devilb2103 @danny-avila 
I implemented the suggested changes in #10676 since I couldn't commit to his branch. 
I also added geminiAPI support and some other small improvements.

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 commented on api/app/clients/tools/manifest.json:
```diff
@@ -179,5 +179,12 @@
+  {
+    "name": "Gemini Image Tools",
+    "pluginKey": "gemini_image_gen",
+    "description": "Generate high-quality images using Google's Gemini 2.5 Flash Image Model",
```
```suggestion
    "description": "Generate high-quality images using Google's Gemini Image Models",
```

devilb2103 commented on api/app/clients/tools/structured/GeminiImageGen.js:
```diff
@@ -0,0 +1,657 @@
+      // Use same pattern as other Google service integrations
+      const credentialsPath =
+        process.env.GOOGLE_SERVICE_KEY_FILE ||
+        path.join(__dirname, '../../../..', 'data', 'auth.json');
```
```suggestion
        process.env.GOOGLE_SERVICE_KEY_FILE || path.join(process.cwd(), 'api', 'data', 'auth.json');
```

devilb2103 commented on api/server/services/ToolService.js:
```diff
@@ -36,6 +36,7 @@ const { recordUsage } = require('~/server/services/Threads');
 const { loadTools } = require('~/app/clients/tools/util');
 const { redactMessage } = require('~/config/parsers');
 const { findPluginAuthsByKeys } = require('~/models');
+// const { GeminiImageGen } = require('~/app/clients/tools/structured/GeminiImageGen');
```
```suggestion

```

devilb2103 commented on api/app/clients/tools/structured/GeminiImageGen.js:
```diff
@@ -0,0 +1,657 @@
+
+  async initializeGeminiVertexAI() {
+    try {
+      // Use same pattern as other Google service integrations
```
```suggestion
      const geminiApiKey = process.env.GEMINI_API_KEY;
      if (geminiApiKey) {
        logger.debug('[GeminiImageGen] Using Gemini API with API key');
        return new GoogleGenAI({ apiKey: geminiApiKey });
      }

      // Fall back to Vertex AI with service account
      logger.debug('[GeminiImageGen] Using Vertex AI with service account');
```

devilb2103 committed:
🔧 update: Upgrade @google/genai to version 1.19.0 and replace PNG icon with SVG for Gemini Image Tools

- Updated the @google/genai dependency in package-lock.json and package.json to version 1.19.0.
- Enhanced the Gemini Image Tools description and changed the icon from PNG to SVG for better scalability and design.
- Refactored GeminiImageGen.js to support provider detection and model ID configuration for improved flexibility in image generation.
- Removed deprecated references and cleaned up the handleTools.js file by eliminating unnecessary parameters.

devilb2103 committed:
✨ feat: Introduce createImageToolContext for Image Tools

- Added a new utility function `createImageToolContext` to streamline the creation of context strings for image generation and editing tools.
- Refactored `handleTools.js` to utilize the new function, improving code readability and maintainability.
- Created a new file `imageContext.ts` to define the interface and implementation for the image tool context functionality.
- Updated the index file to export the new image context utilities.

devilb2103 committed:
.

devilb2103 committed:
🔧 refactor: Migrate GeminiImageGen tool from CommonJS to ts and update related files

- Migrated `GeminiImageGen` tool implementationfrom CommonJS to TS.
- Updated the `index.js` `GeminiImageGen` references.
- Enhanced the README documentation for the Gemini Image Generation Tool, reflecting the new structure and features.
- Adjusted the `handleTools.js` and `tools.js` files to ensure proper integration of remaining tools and maintain functionality.
- Added constants and types for the Gemini tool to improve code organization and clarity.

devilb2103 committed:
✨ feat: Extend image generation support for Gemini tool

- Added support for the `gemini_image_gen` tool in the message rendering logic.
- Updated progress text handling for `gemini_image_gen` to provide detailed status messages during image creation.
- Enhanced user feedback with localized messages reflecting the progress of image generation.

devilb2103 committed:
🗑️ chore: Remove outdated README for Gemini Image Generation Tool

- Deleted the README-GeminiNanoBanana.md file as it is no longer relevant to the current implementation of the Gemini image generation tool.

devilb2103 commented on api/app/clients/tools/structured/README-GeminiNanoBanana.md:
I have removed the md file. I'll add on to the documentation once I get a review on this pr and everything seems good to go!

devilb2103 committed:
🔧 refactor: Improve provider detection and error handling in GeminiImageGen

- Enhanced auto-detection logic to prefer Vertex AI only if the GOOGLE_SERVICE_KEY_FILE exists and is valid.
- Updated the credentials path to use the current working directory for the default auth.json file.
- Added error handling for malformed JSON in the Google service account credentials file, providing clearer error messages.

devilb2103 commented on packages/api/src/tools/gemini/GeminiImageGen.ts:
```diff
@@ -0,0 +1,727 @@
+      }
+
+      process.env.GOOGLE_APPLICATION_CREDENTIALS = credentialsPath;
+      const serviceKey = JSON.parse(fs.readFileSync(credentialsPath, 'utf8'));
```
addressed

devilb2103 committed:
🔧 refactor: Simplify credentials path handling in GeminiImageGen

- Updated the method for retrieving the credentials file path to streamline the logic and remove unnecessary environment variable checks.
- Adjusted the default path for the auth.json file to reflect the current working directory more accurately.

devilb2103 commented on packages/api/src/tools/gemini/GeminiImageGen.ts:
```diff
@@ -0,0 +1,727 @@
+          const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
+          chunks.push(new Uint8Array(buf));
+        }
+        const buffer = Buffer.concat(chunks);
```
@danny-avila this seems false positive. Size limits already exist, so upload shouldn't be an issue.

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana

devilb2103 committed:
Merge branch 'dev' into Gemini-Nano-Banana
